ryu/d2s.c

```diff
@@ -49,16 +49,77 @@
 #define HAS_64_BIT_INTRINSICS
 #endif
 
+#if defined(_M_IX86)
+#define RYU_DIVCONST 1
+#include <intrin.h>
+#endif
+
 #include "ryu/common.h"
 #include "ryu/digit_table.h"
 #include "ryu/d2s.h"
 
+#if RYU_DIVCONST
+// Manually replace division-by-constant by a multiplication and shifts.
+
+// Returns the high 64 bits of the 128-bit product of a and b.
+static inline uint64_t umulh(const uint64_t a, const uint64_t b) {
+//#if defined(_M_ARM)
+//  return _MulUnsignedHigh(a, b);
+//#else
+  const uint32_t aLo = (uint32_t)(a);
+  const uint32_t aHi = (uint32_t)(a >> 32);
+  const uint32_t bLo = (uint32_t)(b);
+  const uint32_t bHi = (uint32_t)(b >> 32);
+
+  const uint32_t b00Hi = (uint32_t)(((uint64_t)aLo * bLo) >> 32);
+  const uint64_t b01 = (uint64_t)aLo * bHi;
+  const uint64_t b10 = (uint64_t)aHi * bLo;
+  const uint64_t b11 = (uint64_t)aHi * bHi;
+
+  const uint64_t mid1 = b10 + b00Hi;
+  const uint32_t mid1Lo = (uint32_t)(mid1);
+  const uint32_t mid1Hi = (uint32_t)(mid1 >> 32);
+
+  const uint64_t mid2 = b01 + mid1Lo;
+  const uint32_t mid2Hi = (uint32_t)(mid2 >> 32);
+
+  return b11 + mid1Hi + mid2Hi;
```

Comment:
This can be done with 4 adds instead of 5: https://stackoverflow.com/questions/28868367/getting-the-high-part-of-64-bit-integer-multiplication/51587262?noredirect=1#comment90287389_51587262